### ISSUE TITLE
Introduce Aggregate Root, Entity, ValueObject, DomainEvent abstracts

### ISSUE THREAD

**Issue by hantsy** [jakartaee/data#374]
In Eric's DDD, a `Repository` is tightly tied with Aggregate Root, and used to maintain the lifecycle of a Aggregate Root. And a Aggregate Root could be consist of Entity, Value Objects, and can handle domain events. 

In a JPA, it is easy to maintain the lifecycle like relation between PurchaseOrder and OrderItems, but switching to Jdbc, it should apple changes to the whole Aggregate when calling a repository save/update etc.

**Comment by gavinking** [comment]
We discussed this issue quickly on the Jakarta Data call yesterday.

It's my opinion&mdash;and I believe that the others agree&mdash;that:

1. Jakarta Data 1.1 provides sufficiently-powerful primitives to express almost any DDD-style Repository that a user could possibly wish to write, but that
2. the specification itself should not directly encode particular patterns or ideologies which are embraced by only a small sliver of the Jakarta community.

The idea of "design patterns" is that they're patterns, not APIs.

On the call we agreed that this issue can now be closed.

**Comment by hantsy** [comment]
We can not avoid the patterns in the real world application development.

If we do not embrace the concept of DDD, here the Data Repository is just a DAO in the classic J2EE Design Patterns.

In an AggregateRoot, we hope can maintain the data easily in a root graph.

**Comment by gavinking** [comment]
> We cannot avoid the patterns in the real world application development.

I mean, sure, but if a pattern requires cannot be implemented without a the use of a specific purpose-designed framework, then it's not a very good pattern.

> If we do not embrace the concept of DDD, here the Data Repository is just a DAO in the classic J2EE Design Patterns.

OK, if you want to use the "Repository" to mean "Repository in DDD", I can't stop you. *I'm* not going to use the word like that, but nothing requires us to agree on definitions of words. 

So yes, concretely, I agree that "Repository" and "DAO" are in fact synonyms.

> In an AggregateRoot, we hope can maintain the data easily in a root graph.

Nothing is stopping you from doing that.

But, just in case you're interested, here's an explanation as to why "aggregates" are a terrible idea when working with relational (as opposed to document or hierarchical) databases.

<img width="835" alt="Image" src="https://github.com/user-attachments/assets/666daec3-29d6-4c2f-afe1-667d91b0828a" />

**Comment by gavinking** [comment]
Of course, @hantsy, if you can identify a concrete problem that makes it hard to implement the aggregate pattern using Jakarta Data, then please let us know what it is.

I assert that there is no such concrete problem, and that it's trivially easy to implement aggregates with Jakarta Data. 

And I have the following very good reason for believing that: we know for sure that Jakarta Data works well with _document databases_, and so we can directly infer that it works just as well for someone who wants to simulate a document database using a relational database. And that's all the "aggregate" pattern is: it's an attempt to shoehorn a tree-like data model onto relational tables.

**Comment by gavinking** [comment]
So this issue has been open for a long time now, with no proposal for any _concrete_ changes to Jakarta Data. 

> On the call we agreed that this issue can now be closed.

I'm going take the liberty of closing it, since I'm not aware of any concrete limitation which would prevent someone from implementation DDD patterns in Jakarta Data. Quite to the contrary, I think Jakarta Data is an excellent solution for people doing DDD.
